code/datums/components/storage.dm

```diff
@@ -0,0 +1,335 @@
+	for (var/obj/O in source.contents)
+		if (O.move_triggered)
+			O.move_trigger(user, kindof)
+
+// user uses source, which is in their active hand
+/datum/component/storage/proc/attack_self(source, mob/user)
+	src.handle_storage(user)
+
+// user attacks M with source
+/datum/component/storage/proc/attack(source, mob/M, mob/user)
+	if (surgeryCheck(M, user))
+		insertChestItem(M, user)
+		return
+
+// after user attacks O with source
+/datum/component/storage/proc/after_attack(atom/source, obj/O, mob/user)
+	if (O in source.contents)
+		user.drop_item()
+		SPAWN_DBG(1 DECI SECOND)
+			O.attack_hand(user)
+
+// after source is dropped, also used as convenience hud update function
+/datum/component/storage/proc/on_dropped()
+	if (hud)
+		hud.update()
+
+// add text to passed tooltip desc
+/datum/component/storage/proc/build_tooltip(atom/source, list/desc)
+	desc += "<br>Holding [get_contents_len(source)]/[slots] objects"
+
+// get contents; filter for items and out for grabs
+/datum/component/storage/proc/get_contents(atom/source, list/cont)
+	for (var/obj/item/I in source.contents)
+		if(!istype(I, /obj/item/grab))
+			cont += I
+
+/datum/component/storage/proc/get_all_contents(atom/source, list/cont)
+	var/list/C = list()
+	src.get_contents(source, C)
+	cont += C
+	for (var/obj/item/I in C)
+		var/list/cont2 = list()
+		SEND_SIGNAL(I, COMSIG_STORAGE_GET_CONTENTS, cont2)
+		cont += cont2
+
+// find a certain type in contents
+/datum/component/storage/proc/find_type(atom/source, item_type, list/found)
+	for (var/A in source.contents)
+		if (istype(A, item_type))
+			found += A
+			return COMSIG_RETURN_SUCCESS
+
+// remove I from source hud and transfer it to target (if provided; may not be provided if I's loc is set again right after)
+/datum/component/storage/proc/transfer_item(atom/source, obj/item/I, var/target)
+	if (target)
+		I.set_loc(target)
+	hud.remove_item(I)
+	return COMSIG_RETURN_SUCCESS
+
+/datum/component/storage/proc/can_fit(source, obj/item/I)
+	if (src.max_wclass <= I.w_class)
+		return COMSIG_RETURN_FAILURE
+	return COMSIG_RETURN_SUCCESS
+
+/datum/component/storage/proc/emp_act(atom/source)
+	if (source.contents.len)
+		for (var/atom/A in source.contents)
+			if (isitem(A))
+				var/obj/item/I = A
+				I.emp_act()
+				SEND_SIGNAL(I, COMSIG_MOVABLE_EMP_ACT)
+	return
+
+// convenience function
+/datum/component/storage/proc/get_contents_len(atom/source)
+	var/items = 0
+	for (var/obj/item/I in source.contents)
+		if(!istype(I, /obj/item/grab))
+			items++
+	return items
+
+// let user add I to source; does not have checks and should be internal
+/datum/component/storage/proc/add_item(atom/source, obj/item/I, mob/user)
+	user.u_equip(I)
+	I.dropped(user)
+	I.set_loc(source)
+	hud.add_item(I)
+	return COMSIG_RETURN_SUCCESS
+
+// handles user opening the source and displaying the hud
+/datum/component/storage/proc/handle_storage(mob/user)
+	var/atom/A = src.parent
```

Comment:
Should I just pass `atom/source` instead of looking at `src.parent`, for consistency?